src/python/pants/core/util_rules/environments.py

```diff
@@ -374,6 +374,41 @@ class RemoteEnvironmentTarget(Target):
 # -------------------------------------------------------------------------------------------
 
 
+@rule_helper
+async def _warn_on_non_local_environments(specified_targets: Iterable[Target], source: str) -> None:
+    """Raise a warning when the user runs a local-only operation against a target that expects a
+    non-local environment.
+
+    The `source` will be used to explain which goal is causing the issue.
+    """
+
+    env_names = [
+        (target, target[EnvironmentField].value)
+        for target in specified_targets
+        if target.has_field(EnvironmentField)
+    ]
+
+    env_tgts = await MultiGet(
+        Get(
+            EnvironmentTarget,
+            EnvironmentNameRequest(
+                name,
+                description_of_origin=f"the `environment` field of the target at `{target.address}`",
+            ),
+        )
+        for target, name in env_names
+    )
+
+    for (tgt, env_name), env_tgt in zip(env_names, env_tgts):
+        if env_tgt.val is not None and not isinstance(env_tgt.val, LocalEnvironmentTarget):
+            logger.warning(
+                f"{source.capitalize()} was called with the target `{tgt.address}`, which "
+                f"specifies the environment `{env_name}`, which is a `{env_tgt.val.alias}`. "
+                f"{source.capitalize()} only runs in the local environment. You may experience "
+                "unexpected behavior."
+            )
```

Comment:
This will result in possibly hundreds of warnings when you run `./pants export ::` and `./pants repl ::` - super noisy. I think for `export`, we need to be more clever. E.g. aggregating into a single log.

`run` is fine because we validate it's only one target

Maybe only implement `run` for now, and in a dedicated follow up implement `repl` and `export` after designing what the batched message should look like? Our error messages for Python resolves could be helpful inspiration

Reply:
Could do a `itertools.groupby` on `env_name`, as that should be a small number, and present per environment the first few targets, and in case of many just `... and X more`.

Reply:
@Eric-Arellano Thanks for flagging this, @kaos offers a good solution, so I'll sort that out there.

Reply:
@kaos @Eric-Arellano take another look. It's a bit more complicated, but seems right?